File: server/proto/edid.go
package proto

type EdidState string

const (
	EdidStateSuccess       EdidState = "success"
	EdidStateUnverified    EdidState = "unverified"
	EdidStateNeedsRecovery EdidState = "needs_recovery"
	EdidStateChipRefused   EdidState = "chip_refused"
	EdidStateBusContention EdidState = "bus_contention"
	EdidStatePreflight     EdidState = "preflight"
	EdidStateInvalidInput  EdidState = "invalid_input"
	EdidStateTimeout       EdidState = "timeout"
	EdidStateGeneric       EdidState = "generic"
)

type ApplyEdidReq struct {
	Profile string `validate:"omitempty"` // sha256 id of a shipped profile
	Data    string `validate:"omitempty"` // base64 blob, exclusive with Profile
}

type DecodeEdidReq struct {
	Data string `validate:"omitempty"`
}

type RestoreEdidReq struct {
	Source string `validate:"required,oneof=factory history"`
	ID     string `validate:"omitempty"` // empty means the most recent entry
}

type GetEdidBackupReq struct {
	ID string `form:"id" validate:"required"`
}

type EdidSummary struct {
	SHA256        string `json:"sha256"`
	Manufacturer  string `json:"manufacturer"`
	Model         string `json:"model"`
	ProductCode   uint16 `json:"productCode"`
	Serial        uint32 `json:"serial"`
	Week          uint8  `json:"week"`
	Year          int    `json:"year"`
	Version       string `json:"version"`
	PreferredMode string `json:"preferredMode"`
	PixelClockKHz int    `json:"pixelClockKhz"`
	Extensions    uint8  `json:"extensions"`
	Audio         bool   `json:"audio"`
}

type EdidPreflight struct {
	Chip               string `json:"chip"`
	Product            string `json:"product"`
	Supported          bool   `json:"supported"`
	RequiresPowerCycle bool   `json:"requiresPowerCycle"`
	ToolAvailable      bool   `json:"toolAvailable"`
	Reason             string `json:"reason,omitempty"`
}

type EdidBackup struct {
	ID        string `json:"id"`
	SHA256    string `json:"sha256"`
	AppliedAt string `json:"appliedAt"`
	Size      int    `json:"size"`
}

type GetEdidRsp struct {
	Active              *EdidSummary  `json:"active"`
	Source              string        `json:"source,omitempty"`
	AppliedAt           string        `json:"appliedAt,omitempty"`
	UnverifiedSinceBoot bool          `json:"unverifiedSinceBoot"`
	Preflight           EdidPreflight `json:"preflight"`
	Backups             []EdidBackup  `json:"backups"`
	FactoryAvailable    bool          `json:"factoryAvailable"`
}

type EdidProfile struct {
	ID            string `json:"id"`
	Manufacturer  string `json:"manufacturer"`
	Model         string `json:"model"`
	PreferredMode string `json:"preferredMode"`
	Source        string `json:"source"`
}

type GetEdidProfilesRsp struct {
	Profiles []EdidProfile `json:"profiles"`
}

type DecodeEdidRsp struct {
	Summary EdidSummary `json:"summary"`
	Detail  any         `json:"detail"`
}

type ApplyEdidRsp struct {
	State              EdidState    `json:"state"`
	Verified           bool         `json:"verified"`
	Retryable          bool         `json:"retryable"`
	RequiresPowerCycle bool         `json:"requiresPowerCycle"`
	Message            string       `json:"message"`
	Summary            *EdidSummary `json:"summary,omitempty"`
	WrittenHex         string       `json:"writtenHex,omitempty"` // needs_recovery only
	ReadHex            string       `json:"readHex,omitempty"`    // needs_recovery only
}

File: web/src/api/edid.ts
import { http } from '@/lib/http.ts';

// a flash holds the i2c bus far longer than a normal request, so it gets its own timeout
const applyTimeout = 180 * 1000;

export type EdidSummary = {
  sha256: string;
  manufacturer: string;
  model: string;
  productCode: number;
  serial: number;
  week: number;
  year: number;
  version: string;
  preferredMode: string;
  pixelClockKhz: number;
  extensions: number;
  audio: boolean;
};

export type EdidPreflight = {
  chip: string;
  product: string;
  supported: boolean;
  requiresPowerCycle: boolean;
  toolAvailable: boolean;
  reason?: string;
};

export type EdidBackup = {
  id: string;
  sha256: string;
  appliedAt: string;
  size: number;
};

export type EdidStatus = {
  active?: EdidSummary;
  source?: string;
  appliedAt?: string;
  unverifiedSinceBoot: boolean;
  preflight: EdidPreflight;
  backups: EdidBackup[];
  factoryAvailable: boolean;
};

export type EdidProfile = {
  id: string;
  manufacturer: string;
  model: string;
  preferredMode: string;
  source: string;
};

export type EdidResult = {
  state: string;
  verified: boolean;
  retryable: boolean;
  requiresPowerCycle: boolean;
  message: string;
  summary?: EdidSummary;
};

// get the active edid, the preflight and the backup history
export function getEdid() {
  return http.get('/api/vm/edid');
}

// get the shipped profile library
export function getProfiles() {
  return http.get('/api/vm/edid/profiles');
}

// decode a blob without touching the chip
export function decode(data: string) {
  return http.post('/api/vm/edid/decode', { data });
}

// flash a shipped profile or an uploaded blob
export function apply(profile: string, data: string) {
  return http.post('/api/vm/edid/apply', { profile, data }, { timeout: applyTimeout });
}

// download the bytes of the last verified flash
export function downloadEdid() {
  return http
    .request({ method: 'get', url: '/api/vm/edid/download', responseType: 'blob' })
    .then((rsp) => save(rsp as unknown as Blob, 'nanokvm-edid.bin'));
}

// download one history entry
export function downloadBackup(id: string) {
  return http
    .request({ method: 'get', url: '/api/vm/edid/backup', params: { id }, responseType: 'blob' })
    .then((rsp) => save(rsp as unknown as Blob, `nanokvm-edid-${id}.bin`));
}

// read a file as the base64 the decode and apply endpoints take
export function toBase64(file: File) {
  return file.arrayBuffer().then((buffer) => {
    const bytes = new Uint8Array(buffer);

    let binary = '';
    for (let i = 0; i < bytes.length; i++) {
      binary += String.fromCharCode(bytes[i]);
    }

    return btoa(binary);
  });
}

// the download routes answer with raw bytes, or with the usual json envelope when they fail
function save(blob: Blob, name: string) {
  if (blob.type.includes('json')) {
    return blob.text().then((text) => {
      throw new Error(JSON.parse(text)?.msg || 'download failed');
    });
  }

  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = name;

  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  URL.revokeObjectURL(url);

  return Promise.resolve();
}
